internal/generator/ecc/template/sswu-fp.go.tmpl

```diff
@@ -0,0 +1,397 @@
+// if not, we get sqrt(Z * u / v). Recall that Z is non-residue
+// The return value is undefined for u = 0
+// Taken from https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/13/ F.2.1.1. for any field
+// The main idea is that since the computation of the square root involves taking large powers of u/v, the inversion of v can be avoided
+func sqrtRatio(z *{{.CoordType}}, u *{{.CoordType}}, v *{{.CoordType}}) bool {
+   tv1 := {{.CoordType}} { {{words64 (index .SqrtRatioParams 5) }} }
+
+   var exp big.Int
+   exp.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 3) ) }} })
+   var tv2, tv3, tv4, tv5 {{.CoordType}}
+   tv2.Exp(*v, &exp)
+   tv3.Mul(&tv2, &tv2)
+   tv3.Mul(&tv3, v)
+
+   // line 5
+   tv5.Mul(u, &tv3)
+
+   exp.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 2) ) }} })
+   tv5.Exp(tv5, &exp)
+   tv5.Mul(&tv5, &tv2)
+   tv2.Mul(&tv5, v)
+   tv3.Mul(&tv5, u)
+
+   // line 10
+   tv4.Mul(&tv3, &tv2)
+   exp.SetBytes([]byte { {{ printList (bytes (index .SqrtRatioParams 4) ) }} })
+   tv5.Exp(tv4, &exp)
+
+   isQr := tv5.IsOne()
+
+   tv2.Mul(&tv3, &{{.CoordType}} { {{words64 (index .SqrtRatioParams 6) }} })
+   tv5.Mul(&tv4, &tv1)
+
+   // line 15
+
+   if !isQr {
+      tv3, tv4 = tv2, tv5
+   }
+
+   exp.Lsh( big.NewInt(1), {{ index .SqrtRatioParams 0 }} - 2)
+   for i := {{ index .SqrtRatioParams 0 }}; i >= 2; i -- {
+      //line 20
+      tv5.Exp(tv4, &exp)
+      e1 := tv5.IsOne()
+
+      tv2.Mul(&tv3, &tv1)
+      tv1.Mul(&tv1, &tv1)
+      tv5.Mul(&tv4, &tv1)
+
+      if !e1 {
+         tv3, tv4 = tv2, tv5
+      }
+
+      exp.Rsh(&exp,1)
+   }
+
+   *z = tv3
+   return isQr
+}
+{{ end }}
+
+//TODO: Use addchain
+//TODO: Might duplicate functionality from mulByConst functions
+// mulByZ multiplies x by {{.Z}} and stores the result in z
+func mulByZ(z *{{.CoordType}}, x *{{.CoordType}}) {
+
+    {{ $ZBitsHi2Lo := reverse (bits .Z) }}
+    {{ $op := "Add"}}
+    {{ if lt .Z 0 }}
+        {{ $op = "Sub" }}
+        var res {{.CoordType}}
+        res.Neg(x)
+    {{ end }}
+    {{ if gt .Z 0 }}
+        res := *x
+    {{ end }}
+    res.Double(&res)
+
+    {{- range $bit := noFirst (noLast $ZBitsHi2Lo) }}
+        {{ if $bit }}
+            res.{{$op}}(&res, x)
+        {{ end }}
+        res.Double(&res)
+    {{- end }}
+
+    {{ if last $ZBitsHi2Lo }}
+        res.{{$op}}(&res, x)
+    {{ end }}
+
+    *z = res
+}
+
+// From https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/13/ Pg 80
+func sswuMapG1(u *{{.CoordType}}) G1Affine {
+
+	var tv1 {{.CoordType}}
+	tv1.Square(u)
+
+	//mul tv1 by Z
+	mulByZ(&tv1, &tv1)
+
+	var tv2 {{.CoordType}}
+	tv2.Square(&tv1)
+	tv2.Add(&tv2, &tv1)
+
+	var tv3 {{.CoordType}}
+	//Standard doc line 5
+	var tv4 {{.CoordType}}
+	tv4.SetOne()
+	tv3.Add(&tv2, &tv4)
+	tv3.Mul(&tv3, &{{.CoordType}}{ {{ words64 .B }} })
+
+	tv4 = {{.CoordType}}{ {{ words64 .A }} }
+	//TODO: Std doc uses conditional move. If-then-else good enough here?
```

Comment:
side remark; if we want constant time, in many cases, if / else won't be enough if the branch execution time are different --> 
it is cleaner to execute both branches, and then have a function that conditionally move the result in another variable. (assembly with CMOV on x86).

Reply:
additional precisions; IF / ELSE are, to my knowledge, most of the time translated by JMP instructions in assembly.